File: typescript/src/modules/metadata.ts
import { CODE_PATTERN } from '../codes.js';
import type { Store } from '../store.js';
import type { ShortUrl } from '../schemas.js';
import { toShortUrl } from '../serialize.js';
import { NOT_FOUND_BODY } from '../errors.js';

export type MetadataResult =
  | { status: 200; body: ShortUrl }
  | { status: 404; body: typeof NOT_FOUND_BODY };

const NOT_FOUND: MetadataResult = { status: 404, body: NOT_FOUND_BODY };

/**
 * Metadata module — read-only lookup, expired or not. Performs no writes,
 * so it never touches click_count. See blueprint/MODULE_BOUNDARIES.md.
 */
export function getMetadata(store: Store, code: string, baseUrl: string): MetadataResult {
  if (!CODE_PATTERN.test(code)) {
    return NOT_FOUND;
  }
  const record = store.get(code);
  if (!record) {
    return NOT_FOUND;
  }
  return { status: 200, body: toShortUrl(record, baseUrl) };
}

File: rust/src/modules/metadata.rs
//! Metadata module — `GET /api/v1/urls/{code}`.
//!
//! Owns: read-only lookup of a short URL record, expired or not — no
//! writes, and deliberately no expiration check
//! (`blueprint/MODULE_BOUNDARIES.md`, "Metadata"). Governed by
//! `blueprint/specs/metadata.md`.

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;

use crate::app::AppState;
use crate::codes::is_valid_code_shape;
use crate::errors::not_found;
use crate::schemas::ShortUrl;

pub async fn get_metadata(State(state): State<AppState>, Path(code): Path<String>) -> Response {
    if !is_valid_code_shape(&code) {
        return not_found();
    }

    let Some(record) = state.store.get(&code) else {
        return not_found();
    };

    (StatusCode::OK, Json(ShortUrl {
        code: record.code.clone(),
        original_url: record.original_url,
        short_url: format!("{}/{}", state.base_url, record.code),
        click_count: record.click_count,
        created_at: record.created_at,
        expires_at: record.expires_at,
    }))
        .into_response()
}
